### ISSUE TITLE
Improve ability to monitor pending queries

### ISSUE THREAD

**Issue by ThomWright** [brianc/node-postgres#2402]
Hi,

I'm trying to find a way to improve our monitoring around the connection pool use, specifically around pending queries.

We're using Prometheus to monitor our services. Prometheus works by 'sampling' a service, say every 10 seconds. The service will return some metrics, each of which is normally a counter (only goes up), a gauge (goes up or down) or a histogram (of e.g. durations).

As an example, monitoring of HTTP requests would normally be done with either a counter (if you're measuring throughput), or a histogram (if you're also measuring latency). When using a counter, you increment the counter every time a new request begins. So, for example, the first time a service is sampled it might return a value of 0 if no requests have come in yet. Next time, if there have been 5 requests, it would return 5, even if they have all finished.

What you wouldn't want to do is simply return how many requests are _currently in flight_. This wouldn't give a very good indication of throughput. You won't see any spikes in requests in between samples, for example.

Currently, as far as I can see, the only way to get information about pending queries is to use `Pool.waitingCount`. This is OK, but suffers from the problem mentioned above. What I would like is similar to what I describe above: keep a count of the number of queries which have been added to the pending queue, so I can see how frequently this happens. I think this is a more effective way of seeing if the pool is nearing capacity and potentially blocking queries from running.

In fact, what would be even better is to have a histogram of _how long queries spend in the queue_.

Does this make sense, and does it sound sensible? Is there another way to do what I want, or is there something else I could/should be measuring instead? Is it just a bad idea?

My proposed solution to the problem is to add some events to the pool, e.g.:

- `queue_push`
- `queue_pop`

Ideas for better names very much welcome! What would be useful if these events gave access to the `PendingItem` (or something) which would allow me to identify the corresponding push/pop events for a single query to measure the duration between them.

Thoughts?

Thanks.

**Comment by brianc** [comment]
It sounds like a good idea, though I'm wondering if it could be done outside the library itself...for example, wrapping the pool w/ a decorator or something that tracks this?

**Comment by ThomWright** [comment]
Agreed that building this on top of the library could be a preferable solution.

That said, I'm not sure how it would be possible. At least not in a way which doesn't monkey patch internal details of the pool (e.g. `_pendingQueue`).

Admittedly I haven't thought about it in much depth!

**Comment by SabbirAHmed3** [comment]
Hi

**Comment by christiangalsterer** [comment]
If you are using pg-promise as a wrapper around node-postgres you might be interested in the new pg-promise-prometheus-exporter available at https://www.npmjs.com/package/@christiangalsterer/pg-promise-prometheus-exporter
